### ISSUE TITLE
Support or add an option for a path.

### ISSUE THREAD

**Issue by s7x** [hakluke/hakoriginfinder#7]
Hello hakluke!

I found myself missing some hits with a very specific case that would need a new parameter and/or a support for path in the target website to uncover, directly to the `-h` flag:

Suppose the main host you want to target has a redirect at it's **/** (root page) to **/index.html** for example, how does the tool handle such thing?
Here is an example:

1 - I shorten an URL of an interact-sh instance:

![image](https://github.com/hakluke/hakoriginfinder/assets/33422344/076bd0d6-f36d-448a-b5c5-6711a157b6bc)

2 - I watch the instance in my term for any new hit:

![image](https://github.com/hakluke/hakoriginfinder/assets/33422344/ed7b9ab5-b1b1-425b-a52c-4e6d91d909af)

3 - I fire up the tool with a random IP to test:

![image](https://github.com/hakluke/hakoriginfinder/assets/33422344/3abaf3d3-916e-4331-bad6-f986a783de5d)

4 - No interaction is made to the interact-sh instance:

![image](https://github.com/hakluke/hakoriginfinder/assets/33422344/54785af8-3a42-40a5-bf7a-d2aa78f20f68)

---

So in case of a website redirecting it's **/** to for example **/index.html**, the fact we cannot yet specify a path with an argument/parameter and/or directly to the `-h` makes this situation complex as **hakoriginfinder** is not "able" to properly get an initial "idea" of the page/host.

What do you think?

**TL/DR**: we miss some hits in case of 301/302 redirects because the main site is not 200' on "/".

**Comment by hakluke** [comment]
🤔 Hmmm yeah I can see why this doesn't work. Honestly I'm unlikely to have
the time to get to this any time soon, but I'll accept a good PR!

Luke

**Comment by modrobert** [comment]
@s7x 

I tried with https://tinyurl.com instead of shorturl.at because shorturl.at seems to block interactsh-client hostnames now. Anyway, it seems to work, although you will get the shortened URL path loaded through the redirect which might not be as expected. Not sure if you used the URL shortener only as an example to point out path support is missing, or if you really want the path to be replaced during a redirect (301), if so  I might be able to fix that as well.

To clarify with example output, it works like this now:

```
$ echo 1.1.1.1 | ./hakoriginfinder -l 500 -h https://tinyurl.com/mvtj6m8z
MATCH https://1.1.1.1/mvtj6m8z 228
```

EDIT: 

After checking some more, turns out the old code did do redirects (but without path) when handling "-h" hostname but never showed it, and the reason you get a lot of NOMATCH results for supplied IP addresses with "prips 1.1.1.0/24" is due to HTTP to HTTPS redirects which resulted in a unique cloudflare response for the 1.1.1.0/24 range.

After thinking some more, it needs to handle redirects, will push another commit soon.

**Comment by modrobert** [comment]
@s7x

Added handling for redirects as follows:

* The "-h" hostname/URL follows redirects
* IP addresses from stdin (e.g. prips) are not following redirects (anymore, they used to)
* Request header"Host" is only hostname (and port if available) part of the "-h" URL according to standard

Example output:
```
$ prips 1.1.1.0/24 | ./hakoriginfinder -l 500 -h https://tinyurl.com/mvtj6m8z
Redirect 301 to: http://one.one.one.one/index.html
Redirect 301 to: https://one.one.one.one/index.html
Redirect 308 to: https://one.one.one.one/
MATCH https://1.1.1.1/ 228
```
EDIT: The 228 distance above is due to a request header host bug, fixed in later commit.

**Comment by modrobert** [comment]
Fixed a request header Host bug:

```
-                req.Header.Add("Host", u.Host)
+                req.Host = u.Host
```

Example output:
```
$ prips 1.1.1.0/24 | ./hakoriginfinder -l 500 -h https://tinyurl.com/mvtj6m8z
Redirect 301 to: http://one.one.one.one/index.html
Redirect 301 to: https://one.one.one.one/index.html
Redirect 308 to: https://one.one.one.one/
NOMATCH https://1.1.1.2/ 56383
NOMATCH https://1.1.1.3/ 56383
MATCH https://1.1.1.1/ 0
```
Levenshtein distance is now "0" in above match because Host header is set correctly.
